test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -0,0 +1,513 @@
+            callback(undefined, fakeServer);
+          });
+      });
+
+      afterEach(() => {
+        topologySelectServers.restore();
+      });
+
+      for (const testData of specTests[specTestName]) {
+        it(testData.description, async () => {
+          await executeSDAMTest(testData);
+        });
+      }
+    });
+  }
+});
+
+function convertOutcomeEvents(events) {
+  return events.map(event => {
+    const eventType = Object.keys(event)[0];
+    const args = [];
+    for (const key of Object.keys(event[eventType])) {
+      const argument = event[eventType][key];
+      if (argument.servers) {
+        const serverEntries = argument.servers.map(server => [
+          server.address,
+          normalizeServerDescription(server)
+        ]);
+        argument.servers = Object.fromEntries(serverEntries);
+      }
+
+      if (typeof argument.topologyType === 'string') {
+        // Translation for our driver's TopologyDescription class
+        argument.type = argument.topologyType;
+        delete argument.topologyType;
+      }
+
+      args.push(argument);
+    }
+
+    // convert snake case to camelCase with capital first letter
+    let eventClass = eventType.replace(/_\w/g, c => c[1].toUpperCase());
+    eventClass = eventClass.charAt(0).toUpperCase() + eventClass.slice(1);
+    const eventConstructor = SDAM_EVENT_CLASSES[eventClass];
+    expect(eventConstructor).to.be.a('function');
+    const eventInstance = new eventConstructor(...args);
+    return eventInstance;
+  });
+}
+
+// iterates through expectation building a path of keys that should not exist (null), and
+// removes them from the expectation (NOTE: this mutates the expectation)
+function findOmittedFields(expected) {
+  const result = [];
+  for (const key of Object.keys(expected)) {
+    if (expected[key] == null) {
+      result.push(key);
+      delete expected[key];
+    }
+  }
+
+  return result;
+}
+
+function normalizeServerDescription(serverDescription) {
+  if (serverDescription.type === 'PossiblePrimary') {
+    // Some single-threaded drivers care a lot about ordering potential primary
+    // servers, in order to speed up selection. We don't care, so we'll just mark
+    // it as `Unknown`.
+    serverDescription.type = 'Unknown';
+  }
+
+  return serverDescription;
+}
+
+function cloneMap(map) {
+  const result = Object.create(null);
+  for (const key of map.keys()) {
+    result[key] = JSON.parse(JSON.stringify(map.get(key)));
+  }
+
+  return result;
+}
+
+function cloneForCompare(event) {
+  const result = JSON.parse(JSON.stringify(event));
+
+  if (event.previousDescription != null && event.previousDescription.servers != null) {
+    result.previousDescription.servers = cloneMap(event.previousDescription.servers);
+  }
+
+  if (event.newDescription != null && event.newDescription.servers != null) {
+    result.newDescription.servers = cloneMap(event.newDescription.servers);
+  }
+
+  return result;
+}
+
+const SDAM_EVENTS = [
```

Comment:
is this just the union of `HEARTBEAT_EVENTS` and `TOPOLOGY_EVENTS` excluding ERROR, TIMEOUT, and CLOSE? would it be worth defining these in terms of existing constants instead of manually entered strings?

Reply:
done!